src/charm.py

```diff
@@ -95,17 +113,62 @@ def _on_config_changed(self, event: ConfigChangedEvent) -> None:
             self.unit.status = WaitingStatus("Waiting for vault initialization secret")
             event.defer()
             return
+        if not self.unit.is_leader() and len(self._other_peer_unit_addresses()) == 0:
+            self.unit.status = WaitingStatus("Waiting for other units to provide their addresses")
+            event.defer()
+            return
+        self.unit.status = MaintenanceStatus("Preparing vault")
         self._set_pebble_plan()
-        self._patch_storage_ownership()
         vault = Vault(url=self._api_address)
         vault.set_token(token=root_token)
-        vault.wait_for_api_available()
+        if not vault.is_api_available():
+            self.unit.status = WaitingStatus("Waiting for vault to be available")
+            event.defer()
+            return
+        if not vault.is_initialized():
+            self.unit.status = WaitingStatus("Waiting for vault to be initialized")
+            event.defer()
+            return
         if vault.is_sealed():
             vault.unseal(unseal_keys=unseal_keys)
+        self._set_peer_relation_unit_address()
         self.unit.status = ActiveStatus()
 
+    def _on_peer_relation_created(self, event: RelationJoinedEvent) -> None:
+        """Handle relation-joined event for the replicas relation."""
+        self._set_peer_relation_unit_address()
+
+    def _on_remove(self, event: RemoveEvent):
+        """Handler triggered when the charm is removed.
+
+        Removes the vault service and the raft data and removes the node from the raft cluster.
+        """
+        if not self._container.can_connect():
+            return
+        root_token, unseal_keys = self._get_initialization_secret_from_peer_relation()
+        if root_token:
+            vault = Vault(url=self._api_address)
+            vault.set_token(token=root_token)
+            if vault.is_api_available() and vault.node_in_raft_peers(node_id=self._node_id):
+                vault.remove_raft_node(node_id=self._node_id)
+        if self._vault_service_is_running():
+            self._container.stop(self._service_name)
+        self._container.remove_path(path=f"{VAULT_RAFT_DATA_PATH}/*", recursive=True)
```

Comment:
Do we really need to remove the data? Couldn't this be left to the admin to remove the storage?

Reply:
If we don't and that a unit gets removed then re-added, this unit number will never spin up properly:

```
2023-09-11T22:23:14.994Z [vault] 2023-09-11T22:23:14.994Z [INFO]  core: vault is unsealed
2023-09-11T22:23:14.994Z [vault] 2023-09-11T22:23:14.994Z [INFO]  storage.raft: entering follower state: follower="Node at 10.1.182.60:8201 [Follower]" leader-address= leader-id=
2023-09-11T22:23:14.994Z [vault] 2023-09-11T22:23:14.994Z [INFO]  core: entering standby mode
2023-09-11T22:23:17.732Z [pebble] GET /v1/plan?format=yaml 122.652µs 200
2023-09-11T22:23:31.570Z [vault] 2023-09-11T22:23:31.570Z [WARN]  storage.raft: heartbeat timeout reached, not part of a stable configuration or a non-voter, not triggering a leader election
```

I'm not sure of the reason yet, I'm looking this up

Reply:
From what I have seen in [this thread](https://github.com/hashicorp/vault/issues/8854), it seems normal practice to clean the data of the node before having it rejoin so that it shows up as a new node. So I'd probably keep the behaviour as is unless you see something else.